src/v/hashing/utils.h

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "utils/functional.h"
+
+#include <absl/hash/hash.h>
+#include <boost/functional/hash.hpp>
+
+#include <functional>
+
+namespace impl {
+template<typename T>
+concept is_absl_hashable = requires(T const& t) {
+    { absl::Hash<T>{}(t) } -> std::convertible_to<std::size_t>;
+};
+
+template<typename T>
+concept is_std_hashable = requires(T const& t) {
+    { std::hash<T>{}(t) } -> std::convertible_to<std::size_t>;
+};
+
+template<typename T>
+concept is_boost_hashable = requires(T const& t) {
+    { boost::hash<T>{}(t) } -> std::convertible_to<std::size_t>;
+};
+
+template<typename T>
+size_t hash_combine(size_t& seed, T const& t) {
+    if constexpr (is_std_hashable<T>) {
+        boost::hash_combine(seed, std::hash<T>{}(t));
+    } else if constexpr (is_absl_hashable<T>) {
+        boost::hash_combine(seed, absl::Hash<T>{}(t));
+    } else if constexpr (is_boost_hashable<T>) {
+        boost::hash_combine(seed, boost::hash<T>{}(t));
+    } else {
+        static_assert(always_false_v<T>, "No hasher found for T");
+    }
+    return seed;
+}
+
+template<typename T>
+size_t hash_combine(size_t& seed, std::optional<T> const& t) {
+    if (!t.has_value()) {
+        // Arbitray value for empty optional, taken from absl::hash
+        boost::hash_combine(seed, 0x12345678);
+        return seed;
+    }
+    return hash_combine<T>(seed, *t);
+}
+
+} // namespace impl
+
+template<typename... Args>
+size_t hash_combine(size_t& seed, Args&&... args) {
+    return (impl::hash_combine(seed, std::forward<Args>(args)), ...);
+}
+
+template<typename... Args>
+size_t hash_combine(Args&&... args) {
```

Comment:
I don't like this overload as it's a bit too easy to drop the seed, which would effectively ignore all the values before it,